File: leetcode/Balanced_Binary_Tree.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isBalanced(TreeNode *root) {
        
        return is_balanced_helper(root) >= 0;
    }
private:
    int is_balanced_helper(TreeNode *root){
        if(!root) return 0;
        if(!root->left && !root->right) return 1;
        
        int l = is_balanced_helper(root->left);
        if(l == -1) return -1;
        
        int r = is_balanced_helper(root->right);
        if(r == -1) return -1;
        
        if(abs(l-r) > 1) return -1;
        else return max(l,r) + 1;
    }
};

File: leetcode/java/BalancedBinaryTree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public boolean isBalanced(TreeNode root) {
        return bh(root) != -1;
    }
    
    private int bh(TreeNode root) {
        if(root == null) return 0;
        
        int l = bh(root.left);
        if(l == -1) return -1;
        
        int r = bh(root.right);
        if(r == -1) return -1;
        
        if(Math.abs(l-r) <= 1) return Math.max(l, r) + 1;
        return -1;
    }
}
